analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs

```diff
@@ -179,12 +188,51 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
 
-        private static ISymbol FirstArgumentSymbol(IInvocationOperationWrapper invocation) =>
-            IArgumentOperationWrapper.FromOperation(invocation.Arguments.First()).Value.TrackedSymbol();
+        private static ISymbol ArgumentSymbol(IInvocationOperationWrapper invocation, int argumentIndex, string parameterName) =>
+            invocation.Arguments[argumentIndex].ToArgument() is var argument
+            && argument.Parameter.Name == parameterName
+                ? argument.Value.TrackedSymbol()
+                : invocation.Arguments.SingleOrDefault(x => x.ToArgument() is var namedArgument && namedArgument.Parameter.Name == parameterName).ToArgument().Value.TrackedSymbol();
+
+        private static RefParamContext FindRefParam(SymbolicContext context) =>
+            BoolRefParamFromArgument(context, KnownType.System_Threading_Monitor, "obj", "Enter", "TryEnter")
+            ?? BoolRefParamFromArgument(context, KnownType.System_Threading_Monitor, "obj", "TryEnter")
+            ?? BoolRefParamFromInstance(context, KnownType.System_Threading_SpinLock, "Enter", "TryEnter")
+            ?? BoolRefParamFromInstance(context, KnownType.System_Threading_SpinLock, "TryEnter");
+
+        private static RefParamContext BoolRefParamFromArgument(SymbolicContext context, KnownType type, string lockObjectParameterName, params string[] methodNames) =>
+            context.Operation.Instance.AsInvocation() is { } invocation
+            && InvocationBoolRefSymbol(invocation, type, methodNames) is { } refParameter
+            && ArgumentSymbol(invocation, 0, lockObjectParameterName) is { } lockObject
+                ? new RefParamContext(context, lockObject, refParameter)
+                : null;
+
+        private static RefParamContext BoolRefParamFromInstance(SymbolicContext context, KnownType type, params string[] methodNames) =>
+            context.Operation.Instance.AsInvocation() is { } invocation
+            && InvocationBoolRefSymbol(invocation, type, methodNames) is { } refParameter
+            && invocation.Instance.TrackedSymbol() is { } lockObject
+                ? new RefParamContext(context, lockObject, refParameter)
+                : null;
+
+        private static ISymbol InvocationBoolRefSymbol(IInvocationOperationWrapper invocation, KnownType type, params string[] methodNames)
+        {
+            if (invocation.TargetMethod.IsAny(type, methodNames))
+            {
+                for (var i = 0; i < invocation.TargetMethod.Parameters.Length; i++)
+                {
+                    if (invocation.TargetMethod.Parameters[i].IsType(KnownType.System_Boolean)
+                        && invocation.TargetMethod.Parameters[i].Name == "lockTaken")
+                    {
+                        return ArgumentSymbol(invocation, i, "lockTaken");
+                    }
+                }
+            }
+            return null;
+        }
```

Comment:
Better way of searching for index.
I'm also testing `RefKind` instead of name, because that's what we care about. 
```
=>
            invocation.TargetMethod.IsAny(type, methodNames)
            && invocation.TargetMethod.Parameters.AsEnumerable().IndexOf(x => x.RefKind == RefKind.Ref && x.IsType(KnownType.System_Boolean)) is var index
            && index >= 0
                ? ArgumentSymbol(invocation, index)
                : null;
```